<div align="center"> 

# Investigating Oil prices and Big Tech. Companies Stock Market 

<p>Between 2001-1-1 - 2023-12-31</p>
</div>

In this project, you will apply the skills you acquired in the course to gather and wrangle real-world data with two datasets of your choice.

You will retrieve and extract the data, assess the data programmatically and visually, accross elements of data quality and structure, and implement a cleaning strategy for the data. You will then store the updated data into your selected database/data store, combine the data, and answer a research question with the datasets.

## Neccessary Imports

In [1]:
## Run this cell to successfully run this notebook
## !pip install yfinance

In [2]:
import requests
import pandas as pd
import yfinance as yf
import requests
import config

## Pipeline
1. Gather Data
2. Asses Data
3. Clean Data
4. Store Data
5. Answering Research Questions

## 1. Gather data

In this section, data using two different (data gathering methods) are extractand and then combined. Two different types of data-gathering methods are use:
- Programmatically downloading files
- Gathering data by accessing APIs

### **1.1.** Problem Statement

In this reserach, the stock prices of major tech companies (such as Apple, Microsoft, Google, Amazon, and META) and oil prices over the period spanning from January 2001 to December 2023 are investigated. By examining historical data for both sectors, the aim is to uncover insights into how changes in oil prices impact the financial performance of big tech companies, and vice versa.
The datasets would include the stock market data for:
- META
- AMazon
- Google
- Apple

### **1.2.** Gathering Data Methods and Code

In this project, two types of data gathering are use: 
- Programmatically downloading files
- Gathering data by accessing APIs

Each of the datasets have at least two variables, and have greater than 500 data samples within.

#### **Dataset 1**: Oils and Petroleum dataset 

**Type:** JSON response

**Method:** The data was gathered using the "by accessing APIs" method from https://api.eia.gov/v2/petroleum/pri/spt/data/

**Dataset variables:**

| Column Name | Description |
|---|---|
| period |  date associated with the data point |
| duoarea | identifier for a specific geographical area |
| area-name | area name |
| product | identifier for the specific petroleum product |
| product-name | name of the product |
| process | identifier for the specific process |
| process-name | The name of the process used |
| series | identifier for the specific data series |
| series-description | descriptive text name or label for the data |

### Gathering by API

In [3]:
## Gather the data using the "Gather data by accessing APIs" method from https://api.eia.gov
api_key = config.EIA_API_KEY
url = f'https://api.eia.gov/v2/petroleum/pri/spt/data/?'

params = {
    'api_key':api_key,
    'start':'2001-01-01', 
    'end':'2023-12-31'
}

response = requests.get(url, params=params)
data = response.json()
oil_data =pd.DataFrame(data['response']['data'])
oil_data.sample(5)

period duoarea      area-name product  \
2498  2002-03-08   Y35NY  NEW YORK CITY   EPMRU   
2375  2006-04-21     RGC             NA   EPMRU   
558   2009-11-06   Y44MB             NA  EPLLPA   
1779  2011-09-23   Y35NY  NEW YORK CITY   EPD2F   
2166  2021-05-28   Y44MB             NA  EPLLPA   

                       product-name process    process-name  \
2498  Conventional Regular Gasoline     PF4  Spot Price FOB   
2375  Conventional Regular Gasoline     PF4  Spot Price FOB   
558                         Propane     PF4  Spot Price FOB   
1779    No 2 Fuel Oil / Heating Oil     PF4  Spot Price FOB   
2166                        Propane     PF4  Spot Price FOB   

                        series  \
2498   EER_EPMRU_PF4_Y35NY_DPG   
2375     EER_EPMRU_PF4_RGC_DPG   
558   EER_EPLLPA_PF4_Y44MB_DPG   
1779   EER_EPD2F_PF4_Y35NY_DPG   
2166  EER_EPLLPA_PF4_Y44MB_DPG   

                                     series-description  
2498  New York Harbor Conventional Gasoline Regular ...  
2375  U.S. Gulf Coast Conventional Gasoline Regular ...  
558   Mont Belvieu, TX Propane Spot Price FOB (Dolla...  
1779  New York Harbor No. 2 Heating Oil Spot Price F...  
2166  Mont Belvieu, TX Propane Spot Price FOB (Dolla...

In [4]:
oil_data.shape

(5000, 9)

> The gathered dataset have 5000 rows and 9 features.

In [5]:
oil_data.to_csv('data/oil_data_api.csv', index=False) ## store to cvs file

> Storing step: the dataset is saved to the local data store directory before moving to the next step.

#### **Dataset 2**: Stocks Dataset

**Type:** Pandas.DataFrame

**Method:** The data was gathered using the "Programmatically downloading files" method from Yahoo Finance using the yfinance library

**Dataset variables:**

| Column Name | Description |
|---|---|
| Date | date associated with the data point |
| Open | The opening price of the stock on that date. |
| High | The highest price reached by the stock on that date. |
| Low | The lowest price reached by the stock on that date. |
| Close | The closing price of the stock on that date. |
| Adj Close | The adjusted closing price, adjusted for dividends and splits. |
| Volume | The number of shares traded on that date. |

### Gathering by Programmatically Downloading

In [6]:
## 2nd data gathering was downloaded programitticallly from Yahoo Finance using the yfinance library
companies = ['AAPL', 'MSFT', 'GOOGL', 'AMZN', 'META']
stocks_data = yf.download(companies, start='2001-01-01', end='2023-12-31')
stocks_data.reset_index(inplace=True)
stocks_data.sample(5)

[*********************100%%**********************]  5 of 5 completed


Price        Date  Adj Close                                               \
Ticker                  AAPL       AMZN      GOOGL        META       MSFT   
4348   2018-04-17  42.197220  75.191498  53.906132  168.314133  89.768417   
945    2004-10-08   0.589703   2.000000   3.442746         NaN  17.493927   
3336   2014-04-09  16.660049  16.590500  28.319498   62.282017  34.208485   
602    2003-05-30   0.270998   1.794500        NaN         NaN  15.250608   
3317   2014-03-13  16.670410  18.575500  29.722145   68.688850  32.027672   

Price       Close                                    ...       Open  \
Ticker       AAPL       AMZN      GOOGL        META  ...       AAPL   
4348    44.560001  75.191498  53.967999  168.660004  ...  44.122501   
945      0.697500   2.000000   3.446697         NaN  ...   0.706429   
3336    18.940001  16.590500  28.351999   62.410000  ...  18.665714   
602      0.320536   1.794500        NaN         NaN  ...   0.323571   
3317    18.951786  18.575500  29.756256   68.830002  ...  19.194286   

Price                                                   Volume             \
Ticker       AMZN      GOOGL        META       MSFT       AAPL       AMZN   
4348    73.114998  53.060001  165.830002  95.000000  106421600  102288000   
945      2.047000   3.471471         NaN  28.100000  359228800  170924000   
3336    16.423500  28.292000   59.630001  39.930000  206169600  101132000   
602      1.780500        NaN         NaN  24.730000  382748800  203806000   
3317    18.830999  30.228979   71.290001  38.419998  257742800  136580000   

Price                                       
Ticker        GOOGL         META      MSFT  
4348     54584000.0   22743000.0  26771000  
945     221390388.0          NaN  49556600  
3336     60632000.0  100215000.0  27398700  
602             NaN          NaN  85096700  
3317     93730176.0   57091000.0  32169700  

[5 rows x 31 columns]

In [7]:
stocks_data.shape

(5785, 31)

> The gathered data has 5785 data points and 31 features.

In [8]:
stocks_data.to_csv('data/stock_data.csv', index=False) ## store to cvs file

> Storing step: the dataset is saved to the local data store directory before moving to the next step.

## 2. Assess data

Assess the data according to data quality and tidiness metrics using the report below.

List **two** data quality issues and **two** tidiness issues. Assess each data issue visually **and** programmatically, then briefly describe the issue you find.  **Make sure you include justifications for the methods you use for the assessment.**


Now that we have gathered the datasets, let's assess the dataset for data quality and structural issues.

Here's a list of the data quality attributes we covered in the course for your reference:

    Completeness
    Validity
    Accuracy
    Consistency
    Uniqueness

### Completeness (Issue #1):

#### Stocks Dataset

Inspecting the dataframe programmatically

In [18]:
stocks_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5785 entries, 0 to 5784
Data columns (total 31 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   (Date, )            5785 non-null   datetime64[ns]
 1   (Adj Close, AAPL)   5785 non-null   float64       
 2   (Adj Close, AMZN)   5785 non-null   float64       
 3   (Adj Close, GOOGL)  4875 non-null   float64       
 4   (Adj Close, META)   2923 non-null   float64       
 5   (Adj Close, MSFT)   5785 non-null   float64       
 6   (Close, AAPL)       5785 non-null   float64       
 7   (Close, AMZN)       5785 non-null   float64       
 8   (Close, GOOGL)      4875 non-null   float64       
 9   (Close, META)       2923 non-null   float64       
 10  (Close, MSFT)       5785 non-null   float64       
 11  (High, AAPL)        5785 non-null   float64       
 12  (High, AMZN)        5785 non-null   float64       
 13  (High, GOOGL)       4875 non-null   float64     

> While the range index is from 0 to 5784, some the columns arn't complete such as (Adj Close, GOOGL), (Adj Close, META), (Close, GOOGL), (Close, META),  (High, GOOGL), (High, META),  (Low, GOOGL),  (Low, META), (Open, GOOGL),  (Open, META),  (Volume, GOOGL), (Volume, META) and (Volume, MSFT).
> These missing data are specifically in data regarding the Google, and the META big tech companies.

In [24]:
stocks_data.isna().sum().sort_values()

Price      Ticker
Date                    0
Volume     AMZN         0
           AAPL         0
Open       MSFT         0
           AMZN         0
           AAPL         0
Low        MSFT         0
           AMZN         0
           AAPL         0
High       AMZN         0
           AAPL         0
           MSFT         0
Adj Close  MSFT         0
           AAPL         0
           AMZN         0
Close      MSFT         0
           AAPL         0
Volume     MSFT         0
Close      AMZN         0
Low        GOOGL      910
Close      GOOGL      910
Open       GOOGL      910
Adj Close  GOOGL      910
High       GOOGL      910
Volume     GOOGL      910
Low        META      2862
Volume     META      2862
Adj Close  META      2862
Open       META      2862
High       META      2862
Close      META      2862
dtype: int64

> There are 910 m,issing records from the Google dataset, and 2862 from the META Dataset.

Inspecting the dataframe visually

#### Oil Dataset

Inspecting the dataframe programmatically

In [25]:
oil_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 9 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   period              5000 non-null   object
 1   duoarea             5000 non-null   object
 2   area-name           5000 non-null   object
 3   product             5000 non-null   object
 4   product-name        5000 non-null   object
 5   process             5000 non-null   object
 6   process-name        5000 non-null   object
 7   series              5000 non-null   object
 8   series-description  5000 non-null   object
dtypes: object(9)
memory usage: 351.7+ KB


> The data have 0 to 4999 indices, and all the columns non-null counts are full. However, it was noticed that there are NA values used to presnet the null values.

In [29]:
oil_data.sample(5)

period duoarea      area-name   product  \
4126  2014-02-21   Y35NY  NEW YORK CITY     EPMRU   
4546  2005-05-06   Y05LA    LOS ANGELES    EPD2DC   
128   2009-08-14   Y35NY  NEW YORK CITY  EPD2DXL0   
1562  2015-11-13     RGC             NA  EPD2DXL0   
2015  2023-04-28     RGC             NA      EPJK   

                           product-name process    process-name  \
4126      Conventional Regular Gasoline     PF4  Spot Price FOB   
4546                        Carb Diesel     PF4  Spot Price FOB   
128   No 2 Diesel Low Sulfur (0-15 ppm)     PF4  Spot Price FOB   
1562  No 2 Diesel Low Sulfur (0-15 ppm)     PF4  Spot Price FOB   
2015             Kerosene-Type Jet Fuel     PF4  Spot Price FOB   

                          series  \
4126     EER_EPMRU_PF4_Y35NY_DPG   
4546    EER_EPD2DC_PF4_Y05LA_DPG   
128   EER_EPD2DXL0_PF4_Y35NY_DPG   
1562    EER_EPD2DXL0_PF4_RGC_DPG   
2015        EER_EPJK_PF4_RGC_DPG   

                                     series-description  
4126  New York Harbor Conventional Gasoline Regular ...  
4546  Los Angeles, CA Ultra-Low Sulfur CARB Diesel S...  
128   New York Harbor Ultra-Low Sulfur No 2 Diesel S...  
1562  U.S. Gulf Coast Ultra-Low Sulfur No 2 Diesel S...  
2015  U.S. Gulf Coast Kerosene-Type Jet Fuel Spot Pr...

In [59]:
nan = ['NA','NaN','-', '0']
oil_data.isin(nan).sum()

period                   0
duoarea                  0
area-name             2754
product                  0
product-name             0
process                  0
process-name             0
series                   0
series-description       0
dtype: int64

> In the oil dataset there appears to be 2754 NA values from the area-name column. The other columns doesn't have NA values.

Inspecting the dataframe visually

Issue and justification: *FILL IN*

### Quality Issue 2:

#### Stocks Dataset

Inspecting the dataframe visually

In [ ]:
## Inspecting the dataframe visually

In [ ]:
stocks_data.dtypes

Price      Ticker
Date                 datetime64[ns]
Adj Close  AAPL             float64
           AMZN             float64
           GOOGL            float64
           META             float64
           MSFT             float64
Close      AAPL             float64
           AMZN             float64
           GOOGL            float64
           META             float64
           MSFT             float64
High       AAPL             float64
           AMZN             float64
           GOOGL            float64
           META             float64
           MSFT             float64
Low        AAPL             float64
           AMZN             float64
           GOOGL            float64
           META             float64
           MSFT             float64
Open       AAPL             float64
           AMZN             float64
           GOOGL            float64
           META             float64
           MSFT             float64
Volume     AAPL               int64
          

Inspecting the dataframe programmatically

In [ ]:
## Inspecting the dataframe programmatically

#### Oil Dataset

Inspecting the dataframe programmatically

Inspecting the dataframe visually

### Tidiness Issue 1:

Inspecting the dataframe programmatically

In [ ]:
## Inspecting the dataframe visually

Inspecting the dataframe visually

In [ ]:
## Inspecting the dataframe programmatically

Issue and justification: *FILL IN*

### Tidiness Issue 2: 

Inspecting the dataframe programmatically

In [ ]:
## Inspecting the dataframe visually

Inspecting the dataframe visually

In [ ]:
## Inspecting the dataframe programmatically

Issue and justification: *FILL IN*

## 3. Clean data
It's time to address the issues found during assessment to clean and polish your data.

Clean the data to solve the 4 issues corresponding to data quality and tidiness found in the assessing step. **Make sure you include justifications for your cleaning decisions.**

After the cleaning for each issue, please use **either** the visually or programatical method to validate the cleaning was succesful.

At this stage, you are also expected to remove variables that are unnecessary for your analysis and combine your datasets. Depending on your datasets, you may choose to perform variable combination and elimination before or after the cleaning stage. Your dataset must have **at least** 4 variables after combining the data.

In [ ]:
# FILL IN - Make copies of the datasets to ensure the raw dataframes 
# are not impacted

### **Quality Issue 1: FILL IN**

In [ ]:
# FILL IN - Apply the cleaning strategy

In [ ]:
# FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Quality Issue 2: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Tidiness Issue 1: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Tidiness Issue 2: FILL IN**

In [ ]:
#FILL IN - Apply the cleaning strategy

In [ ]:
#FILL IN - Validate the cleaning was successful

Justification: *FILL IN*

### **Remove unnecessary variables and combine datasets**

Depending on the datasets, you can also peform the combination before the cleaning steps.

In [ ]:
#FILL IN - Remove unnecessary variables and combine datasets

## 4. Update the data store
In this section, the local database/data store in the directory data are updated with the cleaned data too, with best practices for storing cleaned data:
- Maintained different instances / versions of data (raw and cleaned data)
- Named the dataset files informatively
- Ensure both the raw and cleaned data are saved to the database/data store directory.

> Now the datasets are cleaned they are saved to separete file.

In [ ]:
oil_data.to_csv('data/oil_data_api_cleaned.csv', index=False) ## store to cvs file

In [ ]:
stocks_data.to_csv('data/stock_data_cleaned.csv', index=False) ## store to cvs file

## 5. Answer the research question

### **5.1:** Define and answer the research question 
Going back to the problem statement in step 1, use the cleaned data to answer the question you raised. Produce **at least** two visualizations using the cleaned data and explain how they help you answer the question.

*Research question:* FILL IN from answer to Step 1

In [ ]:
#Visual 1 - FILL IN

*Answer to research question:* FILL IN

In [ ]:
#Visual 2 - FILL IN

*Answer to research question:* FILL IN

### **5.2:** Reflection
In 2-4 sentences, if you had more time to complete the project, what actions would you take? For example, which data quality and structural issues would you look into further, and what research questions would you further explore?

*Answer:* FILL IN